include/envoy/server/lifecycle_notifier.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <functional>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Server {
+
+class ServerLifecycleNotifier {
+public:
+  virtual ~ServerLifecycleNotifier() = default;
+
+  /**
+   * Stages of the envoy server instance lifecycle.
+   */
+  enum class Stage {
+    /**
+     * The server instance main thread has entered the dispatcher loop.
```

Comment:
I think the code guarantees that the startup callbacks will be the first things executed in the main dispatcher loop. Is that correct? I think it might be useful to expand the details a bit in the comments here to cover any subtleties.

Reply:
Based on the implementation of DispatcherImpl::run this will be called just before entering the actual dispatcher loop when [flushing](https://github.com/envoyproxy/envoy/blob/8b713b324c02f31784d4b3603c68b77c9e62b371/source/common/event/dispatcher_impl.cc#L166) all callbacks that were posted before (such as when other modules were constructed). It will probably not be the very first thing invoked but it will be called very soon after starting up.